packages/framework/presence/src/systemWorkspace.ts

```diff
@@ -151,9 +152,20 @@ class SystemWorkspaceImpl implements PresenceStatesInternal, SystemWorkspace {
 			// Check new attendee against audience to see if they're currently connected
 			const isAttendeeConnected = audienceMembers.has(clientConnectionId);
 
-			if (isNew && isAttendeeConnected) {
-				// If the attendee is both new and in audience (i.e. currently connected), emit an attendeeJoined event.
-				postUpdateActions.push(() => this.events.emit("attendeeJoined", attendee));
+			if (isAttendeeConnected) {
+				connectedAttendees.add(attendee);
+				if (attendee.getConnectionStatus() === SessionClientStatus.Disconnected) {
+					attendee.setConnectionId(clientConnectionId);
+				}
+				if (isNew) {
+					// If the attendee is both new and in audience (i.e. currently connected), emit an attendeeJoined event.
+					postUpdateActions.push(() => this.events.emit("attendeeJoined", attendee));
+				}
+			}
+
+			// If the attendee is not in the audience, they are considered disconnected.
+			if (!connectedAttendees.has(attendee)) {
```

Comment:
would this just be `else {`?
and `connectedAttendees` could go away.?

Reply:
I believe we need the set to check if we've seen if the attendee has already been marked as being connected or not. Without the set, we could see an old connection Id, check it against audience and assume the attendee is disconnected but it's current connection Id in audience was already checked.

Reply:
My head is not resolving the cases/logic well. But this is definitely more efficient. Since your current tests pass and none of my local additional tests fail with these changes, then I think we are okay. We should follow up with comments.